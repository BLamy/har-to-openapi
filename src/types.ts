import type { OpenApiSpec } from "@loopback/openapi-v3-types";

export interface HarToOpenAPIConfig {
  /** if true, we'll treat every url as having the same domain, regardless of what its actual domain is
    * the first domain we see is the domain we'll use
    * @defaultValue `false` */
  forceAllRequestsInSameSpec?: boolean;
  /** if true, every path object will have its own servers entry, defining its base path. This is useful when
    * forceAllRequestsInSameSpec is set */
  addServersToPaths?: boolean;
  /** try and guess common auth headers
    * @defaultValue `true` */
  guessAuthenticationHeaders?: boolean;
  /** if the response has this status code, ignore the body */
  ignoreBodiesForStatusCodes?: number[];
  /** whether non standard methods should be allowed (like HTTP MY_CUSTOM_METHOD)
    * @defaultValue `false` */
  relaxedMethods?: boolean;
  /** whether we should try and parse non application/json responses as json
    * @defaultValue `true` */
  relaxedContentTypeJsonParse?: boolean;
  /** a list of tags that match passed on the path, either [match_and_tag] or [match, tag] */
  tags?: ([string] | [string, string] | string)[] | ((url: string) => string | string[] | void);
  /** response mime types to filter for */
  mimeTypes?: string[];
  /** known security headers for this har, to add to security field in openapi (e.g. "X-Auth-Token") */
  securityHeaders?: string[];
  /** Whether to filter out all standard headers from the parameter list in openapi
    * @defaultValue `true` */
  filterStandardHeaders?: boolean;
  /** Whether to log errors to console
    * @defaultValue `false` */
  logErrors?: boolean;
  /** a string, regex, or callback to filter urls for inclusion */
  urlFilter?: string | RegExp | ((url: string) => boolean | Promise<boolean>);
  /** when we encounter a URL, try and parameterize it, such that something like
    * GET /uuids/123e4567-e89b-12d3-a456-426655440000 becomes GET /uuids/{uuid}
    * @defaultValue `false` */
  attemptToParameterizeUrl?: boolean;
  /** when we encounter a path without a response or with a response that does not have 2xx, dont include it
    * @defaultValue `false` */
  dropPathsWithoutSuccessfulResponse?: boolean;
}

type Required<T, K extends keyof T> = T & { [P in K]-?: T[P] };
type WithRequired<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>> & Required<T, K>;

export type InternalConfig = WithRequired<
  HarToOpenAPIConfig,
  | "filterStandardHeaders"
  | "addServersToPaths"
  | "attemptToParameterizeUrl"
  | "relaxedMethods"
  | "guessAuthenticationHeaders"
  | "dropPathsWithoutSuccessfulResponse"
  | "relaxedContentTypeJsonParse"
  | "forceAllRequestsInSameSpec"
>;

export interface HarToOpenAPISpec {
  spec: OpenApiSpec;
  yamlSpec: string;
  domain: string | undefined;
}
