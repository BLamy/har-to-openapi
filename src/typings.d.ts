type JSONSchema = object;

declare module "json-schema-deref-sync" {
  declare function index(schema: JSONSchema, options?: json_schema_deref_sync.Options): JSONSchema;
  interface Options {
    baseFolder?: string;
    failOnMissing?: boolean;
    loaders?: {
      [key: string]: (reference: string, options?: Options) => JSONSchema;
    };
    mergeAdditionalProperties?: boolean;
    removeIds?: boolean;
  }

  export = index;
}
