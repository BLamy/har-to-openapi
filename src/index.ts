import type { OperationObject } from "@loopback/openapi-v3-types";
import { createEmptyApiSpec } from "@loopback/openapi-v3-types";
import type { Entry, Har, QueryString } from "har-format";
import YAML from "js-yaml";
import sortJson from "sort-json";
import { addMethod, addQueryStringParams, addRequestHeaders, getBody, getResponseBody, getSecurity } from "./helpers";
import type { Config, IGenerateSpecResponse, InternalConfig } from "./types";
import type {
  PathItemObject,
  PathsObject,
  SecurityRequirementObject,
  ServerObject,
} from "openapi3-ts/src/model/OpenApi";
import { cloneDeep, groupBy } from "lodash-es";
import { addResponse } from "./utils/baseResponse";
import { isStandardMethod } from "./utils/methods";

const checkPathFromFilter = async (urlPath: string, filter: Config["urlFilter"]) => {
  if (typeof filter === "string") {
    return urlPath.includes(filter);
  }
  if (filter instanceof RegExp) {
    return filter.test(urlPath);
  }
  if (typeof filter === "function") {
    return filter(urlPath);
  }
};

const getConfig = (config?: Config): InternalConfig => {
  const internalConfig = cloneDeep(config || {}) as InternalConfig;
  // set up some defaults
  internalConfig.filterStandardHeaders ??= true;
  internalConfig.relaxedContentTypeJsonParse ??= true;
  internalConfig.forceAllRequestsInSameSpec ??= false;
  internalConfig.relaxedMethods ??= false;
  internalConfig.logErrors ??= false;
  if (internalConfig.securityHeaders) {
    internalConfig.securityHeaders = internalConfig.securityHeaders.map((l) => l.toLowerCase());
  }
  return internalConfig;
};

function tryGetHostname(url: string, logErrors: boolean | undefined, fallback: string): string;
function tryGetHostname(url: string, logErrors: boolean | undefined): string | undefined;
function tryGetHostname(url: string, logErrors?: boolean, fallback?: string): string | undefined {
  try {
    return new URL(url).hostname;
  } catch {
    if (logErrors) {
      console.error(`Error parsing url ${url}`);
    }
  }
  return fallback;
}

const generateSpecs = async <T extends Har>(har: T, config?: Config): Promise<IGenerateSpecResponse[]> => {
  if (!har?.log?.entries?.length) {
    return [];
  }

  // decode base64 now
  har.log.entries.forEach((item) => {
    const response = item.response;
    if (response && response.content?.encoding === "base64") {
      response.content.text = Buffer.from(response.content.text || "", "base64").toString();
      delete response.content.encoding;
    }
  });

  const internalConfig = getConfig(config);
  const {
    ignoreBodiesForStatusCodes,
    mimeTypes,
    securityHeaders,
    filterStandardHeaders,
    forceAllRequestsInSameSpec,
    urlFilter,
    relaxedMethods,
    logErrors,
  } = internalConfig;

  const groupedByHostname = groupBy(har.log.entries, (entry: Entry) => {
    if (forceAllRequestsInSameSpec) {
      return "specs";
    }
    return tryGetHostname(entry.request.url, logErrors);
  });
  const specs: IGenerateSpecResponse[] = [];

  for (const domain in groupedByHostname) {
    try {
      // loop through har entries
      const spec = createEmptyApiSpec();
      spec.info.title = "HarToOpenApi";

      const harEntriesForDomain = groupedByHostname[domain];

      const securitySchemas: SecurityRequirementObject[] = [];
      const firstUrl = harEntriesForDomain[0].request.url;

      for (const item of harEntriesForDomain) {
        try {
          const url = item.request.url;

          if (!url) {
            continue;
          }

          // filter and collapse path urls
          const urlObj = new URL(url);
          const urlPath = urlObj.pathname;

          const queryParams = urlObj.search;

          if (urlFilter) {
            const isValid = await checkPathFromFilter(urlObj.href, urlFilter);
            if (!isValid) {
              continue;
            }
          }
          const mimeType = item.response?.content?.mimeType;
          const isValidMimetype = !mimeTypes || (mimeType && mimeTypes.includes(mimeType));
          if (!isValidMimetype) {
            continue;
          }

          // create method
          const method = item.request.method.toLowerCase();
          // if its not standard and we're not in relaxed mode, skip it
          if (!relaxedMethods && !isStandardMethod(method)) {
            continue;
          }
          // create path if it doesn't exist
          spec.paths[urlPath] ??= { parameters: [] } as PathsObject;
          const path = spec.paths[urlPath] as PathItemObject;

          path[method] ??= addMethod(method, urlObj, config);
          const specMethod = path[method] as OperationObject;
          // generate response
          const status = item.response?.status;
          if (status) {
            specMethod.responses[status] ??= addResponse(status, method);
          }

          const requestHeaders = item.request.headers;
          if (securityHeaders?.length && requestHeaders?.length) {
            const security = getSecurity(requestHeaders, securityHeaders);
            if (security) {
              securitySchemas.push(security);
              specMethod.security = [security];
            }
          }

          // add query string parameters
          if (item.request.queryString?.length) {
            addQueryStringParams(specMethod, item.request.queryString);
          }
          if (queryParams) {
            // try and parse from the url if the har is malformed
            const queryStrings: QueryString[] = [];
            for (const entry of urlObj.searchParams.entries()) {
              queryStrings.push({ name: entry[0], value: entry[1] });
            }
            addQueryStringParams(specMethod, queryStrings);
          }
          if (requestHeaders?.length) {
            addRequestHeaders(specMethod, requestHeaders, filterStandardHeaders);
          }

          const shouldUseRequestAndResponse =
            !ignoreBodiesForStatusCodes || !ignoreBodiesForStatusCodes.includes(status);
          if (shouldUseRequestAndResponse && item.request.postData) {
            specMethod.examples ??= [];
            specMethod.requestBody = await getBody(
              item.request.postData,
              { urlPath, method, examples: specMethod.examples },
              internalConfig,
            );
          }

          if (status && isValidMimetype && shouldUseRequestAndResponse && item.response) {
            specMethod.responseExamples ??= {};
            specMethod.responseExamples[status] ??= [];
            const body = await getResponseBody(
              item.response,
              { urlPath, method, examples: specMethod.responseExamples[status] },
              internalConfig,
            );
            if (body) {
              specMethod.responses[status] = body;
            }
          }
        } catch (e) {
          if (logErrors) {
            console.error(`Error parsing ${item.request}`);
            console.error(e);
          }
          // error parsing one entry, move on
        }
      }
      // If there were no valid paths, bail
      if (!Object.keys(spec.paths).length) {
        continue;
      }

      if (securitySchemas.length) {
        spec.components ||= {};
        spec.components.securitySchemes = {};
        securitySchemas.forEach((schema) => {
          const schemaName = Object.keys(schema)[0];
          spec.components!.securitySchemes![schemaName] = {
            type: "http",
            name: schemaName,
            in: "header",
          };
        });
      }

      // remove the examples that we used to build the superset schemas
      for (const path of Object.values(spec.paths)) {
        for (const method of Object.values<OperationObject>(path)) {
          delete method.responseExamples;
          delete method.examples;
        }
      }
      // sort paths
      spec.paths = sortJson(spec.paths, { depth: 200 });
      const yamlSpec = YAML.dump(spec);
      const labeledDomain = tryGetHostname(firstUrl, logErrors, domain);
      const prefix = firstUrl?.startsWith("https://") ? "https://" : "http://";
      const server: ServerObject = {
        url: `${prefix}${labeledDomain}`,
      };
      spec.servers = [server];
      specs.push({ spec, yamlSpec, domain: labeledDomain });
    } catch (err) {
      if (logErrors) {
        console.error(`Error creating spec for ${domain}`);
      }
    }
  }

  return specs;
};
const generateSpec = async <T extends Har>(har: T, config?: Config): Promise<IGenerateSpecResponse> => {
  const specs = await generateSpecs(har, config);
  if (specs.length) {
    return specs[0];
  }
  const spec = createEmptyApiSpec();
  spec.info.title = "HarToOpenApi - no valid specs found";
  return { spec, yamlSpec: YAML.dump(spec), domain: undefined };
};

export { generateSpec, generateSpecs };
