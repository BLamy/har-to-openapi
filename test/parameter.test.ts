import { describe, test } from "vitest";
import { generateSpec } from "../src";
import { parameterizedUrlHar } from "./test-utils";
import OpenAPISchemaValidator from "openapi-schema-validator";

const har = parameterizedUrlHar();
const validator = new OpenAPISchemaValidator({
  version: 3,
});
describe("Parameterized URLs", async () => {
  test(`Can parse the parameterized url properly`, async ({ expect }) => {
    const data = await generateSpec(har, { filterStandardHeaders: true, attemptToParameterizeUrl: true });
    expect(data).toBeDefined();
    expect(data).toMatchSnapshot();
    const result = validator.validate(data.spec as any);
    expect(result.errors).toHaveLength(0);
  });
});
