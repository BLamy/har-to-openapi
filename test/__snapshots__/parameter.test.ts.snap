// Vitest Snapshot v1

exports[`Parameterized URLs > Can parse the parameterized url properly 1`] = `
{
  "domain": "google.com",
  "spec": {
    "info": {
      "title": "HarToOpenApi",
      "version": "1.0.0",
    },
    "openapi": "3.0.0",
    "paths": {
      "/login/date/{date}": {
        "parameters": [
          {
            "in": "path",
            "name": "date",
            "required": true,
          },
        ],
        "post": {
          "description": "",
          "operationId": "postLoginDate20220202",
          "parameters": [],
          "requestBody": {
            "content": {
              "application/json": {
                "example": {
                  "meta": {},
                  "params": {
                    "test": "true",
                  },
                },
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "meta": {
                      "additionalProperties": false,
                      "title": "Meta",
                      "type": "object",
                    },
                    "params": {
                      "additionalProperties": false,
                      "properties": {
                        "test": {
                          "format": "boolean",
                          "type": "string",
                        },
                      },
                      "required": [],
                      "title": "Params",
                      "type": "object",
                    },
                  },
                  "required": [],
                  "title": "LoginDatePostRequest",
                  "type": "object",
                },
              },
            },
            "required": true,
          },
          "responses": {
            "200": {
              "content": {
                "application/json": {
                  "example": {
                    "error": null,
                    "meta": {},
                    "result": {
                      "email": "luis.tejeda@company.com",
                      "id": 1,
                      "name": "Luis Tejeda Sanchez",
                      "role": "USER",
                      "test": "true",
                    },
                  },
                  "schema": {
                    "additionalProperties": false,
                    "properties": {
                      "error": {
                        "nullable": true,
                      },
                      "meta": {
                        "additionalProperties": false,
                        "title": "Meta",
                        "type": "object",
                      },
                      "result": {
                        "additionalProperties": false,
                        "properties": {
                          "email": {
                            "type": "string",
                          },
                          "id": {
                            "type": "integer",
                          },
                          "name": {
                            "type": "string",
                          },
                          "role": {
                            "type": "string",
                          },
                          "test": {
                            "format": "boolean",
                            "type": "string",
                          },
                        },
                        "required": [],
                        "title": "Result",
                        "type": "object",
                      },
                    },
                    "required": [],
                    "title": "LoginDatePostRequest",
                    "type": "object",
                  },
                },
              },
              "description": "",
            },
          },
        },
      },
      "/login/date/{date}/id/{id}/uuid/{uuid}": {
        "parameters": [
          {
            "in": "path",
            "name": "date",
            "required": true,
          },
          {
            "in": "path",
            "name": "id",
            "required": true,
          },
          {
            "in": "path",
            "name": "uuid",
            "required": true,
          },
        ],
        "post": {
          "description": "",
          "operationId": "postLoginDate20220202Id12345Uuid123E4567E89B12D3A456426655440000",
          "parameters": [],
          "requestBody": {
            "content": {
              "application/json": {
                "example": {
                  "meta": {},
                  "params": {
                    "test": "true",
                  },
                },
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "meta": {
                      "additionalProperties": false,
                      "title": "Meta",
                      "type": "object",
                    },
                    "params": {
                      "additionalProperties": false,
                      "properties": {
                        "test": {
                          "format": "boolean",
                          "type": "string",
                        },
                      },
                      "required": [],
                      "title": "Params",
                      "type": "object",
                    },
                  },
                  "required": [],
                  "title": "LoginDateIDUUIDPostRequest",
                  "type": "object",
                },
              },
            },
            "required": true,
          },
          "responses": {
            "200": {
              "content": {
                "application/json": {
                  "example": {
                    "error": null,
                    "meta": {},
                    "result": {
                      "email": "luis.tejeda@company.com",
                      "id": 1,
                      "name": "Luis Tejeda Sanchez",
                      "role": "USER",
                      "test": "true",
                    },
                  },
                  "schema": {
                    "additionalProperties": false,
                    "properties": {
                      "error": {
                        "nullable": true,
                      },
                      "meta": {
                        "additionalProperties": false,
                        "title": "Meta",
                        "type": "object",
                      },
                      "result": {
                        "additionalProperties": false,
                        "properties": {
                          "email": {
                            "type": "string",
                          },
                          "id": {
                            "type": "integer",
                          },
                          "name": {
                            "type": "string",
                          },
                          "role": {
                            "type": "string",
                          },
                          "test": {
                            "format": "boolean",
                            "type": "string",
                          },
                        },
                        "required": [],
                        "title": "Result",
                        "type": "object",
                      },
                    },
                    "required": [],
                    "title": "LoginDateIDUUIDPostRequest",
                    "type": "object",
                  },
                },
              },
              "description": "",
            },
          },
        },
      },
      "/login/id/{id}": {
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
          },
        ],
        "post": {
          "description": "",
          "operationId": "postLoginId1234567",
          "parameters": [],
          "requestBody": {
            "content": {
              "application/json": {
                "example": {
                  "meta": {},
                  "params": {
                    "test": "true",
                  },
                },
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "meta": {
                      "additionalProperties": false,
                      "title": "Meta",
                      "type": "object",
                    },
                    "params": {
                      "additionalProperties": false,
                      "properties": {
                        "test": {
                          "format": "boolean",
                          "type": "string",
                        },
                      },
                      "required": [],
                      "title": "Params",
                      "type": "object",
                    },
                  },
                  "required": [],
                  "title": "LoginIDPostRequest",
                  "type": "object",
                },
              },
            },
            "required": true,
          },
          "responses": {
            "200": {
              "content": {
                "application/json": {
                  "example": {
                    "error": null,
                    "meta": {},
                    "result": {
                      "email": "luis.tejeda@company.com",
                      "id": 1,
                      "name": "Luis Tejeda Sanchez",
                      "role": "USER",
                      "test": "true",
                    },
                  },
                  "schema": {
                    "additionalProperties": false,
                    "properties": {
                      "error": {
                        "nullable": true,
                      },
                      "meta": {
                        "additionalProperties": false,
                        "title": "Meta",
                        "type": "object",
                      },
                      "result": {
                        "additionalProperties": false,
                        "properties": {
                          "email": {
                            "type": "string",
                          },
                          "id": {
                            "type": "integer",
                          },
                          "name": {
                            "type": "string",
                          },
                          "role": {
                            "type": "string",
                          },
                          "test": {
                            "format": "boolean",
                            "type": "string",
                          },
                        },
                        "required": [],
                        "title": "Result",
                        "type": "object",
                      },
                    },
                    "required": [],
                    "title": "LoginIDPostRequest",
                    "type": "object",
                  },
                },
              },
              "description": "",
            },
          },
        },
      },
      "/login/id/{id}/idTwo/{id1}": {
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
          },
          {
            "in": "path",
            "name": "id1",
            "required": true,
          },
        ],
        "post": {
          "description": "",
          "operationId": "postLoginId1234567IdTwo",
          "parameters": [],
          "requestBody": {
            "content": {
              "application/json": {
                "example": {
                  "meta": {},
                  "params": {
                    "test": "true",
                  },
                },
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "meta": {
                      "additionalProperties": false,
                      "title": "Meta",
                      "type": "object",
                    },
                    "params": {
                      "additionalProperties": false,
                      "properties": {
                        "test": {
                          "format": "boolean",
                          "type": "string",
                        },
                      },
                      "required": [],
                      "title": "Params",
                      "type": "object",
                    },
                  },
                  "required": [],
                  "title": "LoginIDIDTwoId1PostRequest",
                  "type": "object",
                },
              },
            },
            "required": true,
          },
          "responses": {
            "200": {
              "content": {
                "application/json": {
                  "example": {
                    "error": null,
                    "meta": {},
                    "result": {
                      "email": "luis.tejeda@company.com",
                      "id": 1,
                      "name": "Luis Tejeda Sanchez",
                      "role": "USER",
                      "test": "true",
                    },
                  },
                  "schema": {
                    "additionalProperties": false,
                    "properties": {
                      "error": {
                        "nullable": true,
                      },
                      "meta": {
                        "additionalProperties": false,
                        "title": "Meta",
                        "type": "object",
                      },
                      "result": {
                        "additionalProperties": false,
                        "properties": {
                          "email": {
                            "type": "string",
                          },
                          "id": {
                            "type": "integer",
                          },
                          "name": {
                            "type": "string",
                          },
                          "role": {
                            "type": "string",
                          },
                          "test": {
                            "format": "boolean",
                            "type": "string",
                          },
                        },
                        "required": [],
                        "title": "Result",
                        "type": "object",
                      },
                    },
                    "required": [],
                    "title": "LoginIDIDTwoId1PostRequest",
                    "type": "object",
                  },
                },
              },
              "description": "",
            },
          },
        },
      },
      "/login/uuid/{uuid}": {
        "parameters": [
          {
            "in": "path",
            "name": "uuid",
            "required": true,
          },
        ],
        "post": {
          "description": "",
          "operationId": "postLoginUuid123E4567E89B12D3A456426655440000",
          "parameters": [],
          "requestBody": {
            "content": {
              "application/json": {
                "example": {
                  "meta": {},
                  "params": {
                    "test": "true",
                  },
                },
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "meta": {
                      "additionalProperties": false,
                      "title": "Meta",
                      "type": "object",
                    },
                    "params": {
                      "additionalProperties": false,
                      "properties": {
                        "test": {
                          "format": "boolean",
                          "type": "string",
                        },
                      },
                      "required": [],
                      "title": "Params",
                      "type": "object",
                    },
                  },
                  "required": [],
                  "title": "LoginUUIDPostRequest",
                  "type": "object",
                },
              },
            },
            "required": true,
          },
          "responses": {
            "200": {
              "content": {
                "application/json": {
                  "example": {
                    "error": null,
                    "meta": {},
                    "result": {
                      "email": "luis.tejeda@company.com",
                      "id": 1,
                      "name": "Luis Tejeda Sanchez",
                      "role": "USER",
                      "test": "true",
                    },
                  },
                  "schema": {
                    "additionalProperties": false,
                    "properties": {
                      "error": {
                        "nullable": true,
                      },
                      "meta": {
                        "additionalProperties": false,
                        "title": "Meta",
                        "type": "object",
                      },
                      "result": {
                        "additionalProperties": false,
                        "properties": {
                          "email": {
                            "type": "string",
                          },
                          "id": {
                            "type": "integer",
                          },
                          "name": {
                            "type": "string",
                          },
                          "role": {
                            "type": "string",
                          },
                          "test": {
                            "format": "boolean",
                            "type": "string",
                          },
                        },
                        "required": [],
                        "title": "Result",
                        "type": "object",
                      },
                    },
                    "required": [],
                    "title": "LoginUUIDPostRequest",
                    "type": "object",
                  },
                },
              },
              "description": "",
            },
          },
        },
      },
    },
    "servers": [
      {
        "url": "https://google.com",
      },
    ],
  },
  "yamlSpec": "openapi: 3.0.0
info:
  title: HarToOpenApi
  version: 1.0.0
paths:
  /login/date/{date}:
    parameters:
      - in: path
        name: date
        required: true
    post:
      description: ''
      operationId: postLoginDate20220202
      parameters: []
      requestBody:
        content:
          application/json:
            example:
              meta: {}
              params:
                test: 'true'
            schema:
              additionalProperties: false
              properties:
                meta:
                  additionalProperties: false
                  title: Meta
                  type: object
                params:
                  additionalProperties: false
                  properties:
                    test:
                      format: boolean
                      type: string
                  required: []
                  title: Params
                  type: object
              required: []
              title: LoginDatePostRequest
              type: object
        required: true
      responses:
        '200':
          content:
            application/json:
              example:
                error: null
                meta: {}
                result:
                  email: luis.tejeda@company.com
                  id: 1
                  name: Luis Tejeda Sanchez
                  role: USER
                  test: 'true'
              schema:
                additionalProperties: false
                properties:
                  error:
                    nullable: true
                  meta:
                    additionalProperties: false
                    title: Meta
                    type: object
                  result:
                    additionalProperties: false
                    properties:
                      email:
                        type: string
                      id:
                        type: integer
                      name:
                        type: string
                      role:
                        type: string
                      test:
                        format: boolean
                        type: string
                    required: []
                    title: Result
                    type: object
                required: []
                title: LoginDatePostRequest
                type: object
          description: ''
  /login/date/{date}/id/{id}/uuid/{uuid}:
    parameters:
      - in: path
        name: date
        required: true
      - in: path
        name: id
        required: true
      - in: path
        name: uuid
        required: true
    post:
      description: ''
      operationId: postLoginDate20220202Id12345Uuid123E4567E89B12D3A456426655440000
      parameters: []
      requestBody:
        content:
          application/json:
            example:
              meta: {}
              params:
                test: 'true'
            schema:
              additionalProperties: false
              properties:
                meta:
                  additionalProperties: false
                  title: Meta
                  type: object
                params:
                  additionalProperties: false
                  properties:
                    test:
                      format: boolean
                      type: string
                  required: []
                  title: Params
                  type: object
              required: []
              title: LoginDateIDUUIDPostRequest
              type: object
        required: true
      responses:
        '200':
          content:
            application/json:
              example:
                error: null
                meta: {}
                result:
                  email: luis.tejeda@company.com
                  id: 1
                  name: Luis Tejeda Sanchez
                  role: USER
                  test: 'true'
              schema:
                additionalProperties: false
                properties:
                  error:
                    nullable: true
                  meta:
                    additionalProperties: false
                    title: Meta
                    type: object
                  result:
                    additionalProperties: false
                    properties:
                      email:
                        type: string
                      id:
                        type: integer
                      name:
                        type: string
                      role:
                        type: string
                      test:
                        format: boolean
                        type: string
                    required: []
                    title: Result
                    type: object
                required: []
                title: LoginDateIDUUIDPostRequest
                type: object
          description: ''
  /login/id/{id}:
    parameters:
      - in: path
        name: id
        required: true
    post:
      description: ''
      operationId: postLoginId1234567
      parameters: []
      requestBody:
        content:
          application/json:
            example:
              meta: {}
              params:
                test: 'true'
            schema:
              additionalProperties: false
              properties:
                meta:
                  additionalProperties: false
                  title: Meta
                  type: object
                params:
                  additionalProperties: false
                  properties:
                    test:
                      format: boolean
                      type: string
                  required: []
                  title: Params
                  type: object
              required: []
              title: LoginIDPostRequest
              type: object
        required: true
      responses:
        '200':
          content:
            application/json:
              example:
                error: null
                meta: {}
                result:
                  email: luis.tejeda@company.com
                  id: 1
                  name: Luis Tejeda Sanchez
                  role: USER
                  test: 'true'
              schema:
                additionalProperties: false
                properties:
                  error:
                    nullable: true
                  meta:
                    additionalProperties: false
                    title: Meta
                    type: object
                  result:
                    additionalProperties: false
                    properties:
                      email:
                        type: string
                      id:
                        type: integer
                      name:
                        type: string
                      role:
                        type: string
                      test:
                        format: boolean
                        type: string
                    required: []
                    title: Result
                    type: object
                required: []
                title: LoginIDPostRequest
                type: object
          description: ''
  /login/id/{id}/idTwo/{id1}:
    parameters:
      - in: path
        name: id
        required: true
      - in: path
        name: id1
        required: true
    post:
      description: ''
      operationId: postLoginId1234567IdTwo
      parameters: []
      requestBody:
        content:
          application/json:
            example:
              meta: {}
              params:
                test: 'true'
            schema:
              additionalProperties: false
              properties:
                meta:
                  additionalProperties: false
                  title: Meta
                  type: object
                params:
                  additionalProperties: false
                  properties:
                    test:
                      format: boolean
                      type: string
                  required: []
                  title: Params
                  type: object
              required: []
              title: LoginIDIDTwoId1PostRequest
              type: object
        required: true
      responses:
        '200':
          content:
            application/json:
              example:
                error: null
                meta: {}
                result:
                  email: luis.tejeda@company.com
                  id: 1
                  name: Luis Tejeda Sanchez
                  role: USER
                  test: 'true'
              schema:
                additionalProperties: false
                properties:
                  error:
                    nullable: true
                  meta:
                    additionalProperties: false
                    title: Meta
                    type: object
                  result:
                    additionalProperties: false
                    properties:
                      email:
                        type: string
                      id:
                        type: integer
                      name:
                        type: string
                      role:
                        type: string
                      test:
                        format: boolean
                        type: string
                    required: []
                    title: Result
                    type: object
                required: []
                title: LoginIDIDTwoId1PostRequest
                type: object
          description: ''
  /login/uuid/{uuid}:
    parameters:
      - in: path
        name: uuid
        required: true
    post:
      description: ''
      operationId: postLoginUuid123E4567E89B12D3A456426655440000
      parameters: []
      requestBody:
        content:
          application/json:
            example:
              meta: {}
              params:
                test: 'true'
            schema:
              additionalProperties: false
              properties:
                meta:
                  additionalProperties: false
                  title: Meta
                  type: object
                params:
                  additionalProperties: false
                  properties:
                    test:
                      format: boolean
                      type: string
                  required: []
                  title: Params
                  type: object
              required: []
              title: LoginUUIDPostRequest
              type: object
        required: true
      responses:
        '200':
          content:
            application/json:
              example:
                error: null
                meta: {}
                result:
                  email: luis.tejeda@company.com
                  id: 1
                  name: Luis Tejeda Sanchez
                  role: USER
                  test: 'true'
              schema:
                additionalProperties: false
                properties:
                  error:
                    nullable: true
                  meta:
                    additionalProperties: false
                    title: Meta
                    type: object
                  result:
                    additionalProperties: false
                    properties:
                      email:
                        type: string
                      id:
                        type: integer
                      name:
                        type: string
                      role:
                        type: string
                      test:
                        format: boolean
                        type: string
                    required: []
                    title: Result
                    type: object
                required: []
                title: LoginUUIDPostRequest
                type: object
          description: ''
servers:
  - url: /
",
}
`;
